### PR TITLE
chain: Wireup go-owasm with blockchain properly

### LINKED ISSUE #1868: Wireup go-owasm with blockchain properly
Tasks include compilation during new / edit oracle script and put proper gas limit

### PR DESCRIPTION
**Pull request by taobun**
fixed #1868

### CODE REVIEW

**Review by sorawit — CHANGES_REQUESTED** [review]
Let's add validate basic to ensure that you can't create do-not-modify data sources or oracle scripts.

File: chain/x/oracle/common_test.go [hunk 1]
```diff
@@ -69,3 +68,45 @@ func getTestOracleScript() (os types.OracleScript, clear func()) {
...
+	}
+	output, err := ioutil.ReadFile(outputFile.Name())
+	if err != nil {
+		panic(err)
+	}
+	return output
+}
+
+func mustGetOwasmCode(filename string) []byte {
+	absPath, _ := filepath.Abs(fmt.Sprintf("../../pkg/owasm/res/%s", filename))
```
Comment (sorawit) [inline chain/x/oracle/common_test.go]: Let's move out of this legacy `pkg/owasm`. We can create `testfiles` folder in `x/oracle`

File: chain/x/oracle/keeper/common_test.go [hunk 2]
```diff
@@ -66,14 +68,19 @@ func getTestDataSource(executable string) (ds types.DataSource, clear func()) {
 }
 
 func getTestOracleScript() (os types.OracleScript, clear func()) {
-	absPath, _ := filepath.Abs("../../../pkg/owasm/res/beeb.wasm")
-	code, err := ioutil.ReadFile(absPath)
+	absPath, _ := filepath.Abs("../../../pkg/owasm/res/beeb.wat")
```
Comment (sorawit) [inline chain/x/oracle/keeper/common_test.go]: also here

File: chain/x/oracle/types/error.go [hunk 3]
```diff
@@ -41,6 +41,7 @@ var (
 	ErrCreateWithDoNotModify       = sdkerrors.Register(ModuleName, 34, "create with [do-not-modify] content")
 	ErrSelfReferenceAsReporter     = sdkerrors.Register(ModuleName, 35, "cannot reference self as reporter")
 	ErrOBIDecode                   = sdkerrors.Register(ModuleName, 36, "cannot obi decode")
+	ErrCompileFailed               = sdkerrors.Register(ModuleName, 37, "owasm compile failed")
```
Comment (sorawit) [inline chain/x/oracle/types/error.go]: ErrOwasmCompilation ... owasm compilation failed

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/handler_test.go [hunk 4]
```diff
@@ -192,17 +208,34 @@ func TestCreateOracleScriptSuccess(t *testing.T) {
...
+		Owner.Address, name, description, expectFilename, schema, url,
+	)
 	require.Equal(t, expectOracleScript, oracleScript)
 }
+
+func TestCreateOracleScriptFailed(t *testing.T) {
+	_, ctx, keeper := createTestInput()
+	name := "os_1"
+	description := "beeb"
+	code := []byte("non wasm coce")
```
Comment (Benzbeeb) [inline chain/x/oracle/handler_test.go]: non wasm code

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

File: chain/pkg/owasm/res/edited_beeb.wat [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  (type $t0 (func))
+  (type $t1 (func (param i64 i64 i64 i64)))
+  (type $t2 (func (param i64 i64)))
+  (import "env" "ask_external_data" (func $ask_external_data (type $t1)))
+  (import "env" "set_return_data" (func $set_return_data (type $t2)))
+  (func $prepare (export "prepare") (type $t0)
+    (local $l0 i64)
+    i64.const 1
+    i64.const 1
+    i32.const 1024
```
Comment (Benzbeeb) [inline chain/pkg/owasm/res/edited_beeb.wat]: change this line to ``` i64.const 1024```

File: chain/pkg/owasm/res/edited_beeb.wat [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    i64.extend_u/i32
```
Comment (Benzbeeb) [inline chain/pkg/owasm/res/edited_beeb.wat]: remove this line

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

File: chain/pkg/owasm/res/edited_beeb.wat [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    get_local $l0
+    i64.const 4
+    call $ask_external_data
+    i64.const 3
+    i64.const 3
+    get_local $l0
+    i64.const 4
+    call $ask_external_data)
+  (func $execute (export "execute") (type $t0)
+    i32.const 1024
```
Comment (Benzbeeb) [inline chain/pkg/owasm/res/edited_beeb.wat]: ditto

File: chain/pkg/owasm/res/edited_beeb.wat [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    i64.extend_u/i32
```
Comment (Benzbeeb) [inline chain/pkg/owasm/res/edited_beeb.wat]: ditto

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

**Review by Benzbeeb — APPROVED** [review]

File: chain/x/oracle/handler_test.go [hunk 9]
```diff
@@ -327,22 +398,27 @@ func TestEditGzippedOracleScriptSuccess(t *testing.T) {
...
-	msg := types.NewMsgCreateOracleScript(Owner.Address, name, description, code, schema, url, Alice.Address)
+	msg := types.NewMsgCreateOracleScript(
+		Owner.Address, name, description, code, schema, url, Alice.Address,
+	)
 	_, err := oracle.NewHandler(keeper)(ctx, msg)
+	require.NoError(t, err)
+
 	dir := filepath.Join(viper.GetString(cli.HomeFlag), "files")
 	filename := keeper.MustGetOracleScript(ctx, 1).Filename
+	fmt.Println(filename)
```
Comment (sorawit) [inline chain/x/oracle/handler_test.go]: remove please

File: chain/x/oracle/handler_test.go [hunk 10]
```diff
@@ -351,33 +427,41 @@ func TestEditGzippedOracleScriptSuccess(t *testing.T) {
...
+		gzippedCode, newSchema, newURL, Owner.Address,
+	)
 	res, err := oracle.NewHandler(keeper)(ctx, msgEdit)
 	require.NoError(t, err)
 	expectEvents := sdk.Events{
 		sdk.NewEvent(types.EventTypeEditOracleScript, sdk.NewAttribute(types.AttributeKeyID, "1")),
 	}
 	require.Equal(t, expectEvents, res.Events)
 	filename = keeper.MustGetOracleScript(ctx, 1).Filename
+	fmt.Println(filename)
```
Comment (sorawit) [inline chain/x/oracle/handler_test.go]: remove

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**